doc: add guide on maintaining the build files

joyeecheung opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

This documents how to maintain the build files and some commonly-used targets in the Makefile. It also documents about how to update certain targets to reflect the changes on CI (especially the Raspberry Pis and Linux with FIPS).

This is still a WIP, and only talks about the makefile, configure script and the vcbuild.bat script. There are other build files (e.g. the gyp files) that need to be documented later.

Personally I prefer to leave most of the actual documentation in the code, whiling leave some pointers and necessary explanations in the guide so it's easier to get started with maintaining them.

cc @nodejs/build 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
doc, build

joyeecheung commented on doc/guides/maintaining-the-build-files.md:
````diff
@@ -0,0 +1,386 @@
+or `.\configure --help` (Windows). Consider updating this file
+when you need to update the configuration process.
+
+```
````
I am not sure if pasting the help text here is useful..although it does give you an idea about what it does if you don't know what it is.

richardlau replied:
I think the above paragraph is enough. Including the text here will only drift from the actual help text over time.

gibfahn replied:
Agreed with @richardlau , don't think we should duplicate the `./configure --help` output.

joyeecheung commented on doc/guides/maintaining-the-build-files.md:
````diff
@@ -0,0 +1,386 @@
+To see the help text, run `.\vcbuild help`. Consider updating this file
+when you need to update the build and testing process on Windows.
+
+```
````
We should document the similarity/differences between  `vcbuild.bat` and `Makefile`, though not necessarily in this PR (I am not familiar enough with `vcbuild.bat` to be comfortable to write a guide on it anyway)

rvagg replied:
that'd be quite a job for someone! one of it's biggest problems is that it lacks most of the smarts that make brings to the table and we only do a partial job of emulating the bits that are needed

joyeecheung commented on Makefile:
```diff
@@ -390,13 +402,17 @@ CI_JS_SUITES ?= default
 CI_DOC := doctool
 
 # Build and test addons without building anything else
+# The CI runs this on ARM(Raspberry Pis) because the addons and
```
Uh, wait, I don't think this is true..? From what I can tell `node-test-binary-arm` only runs `test-ci-js`, But then I am not sure who uses this target and how are the addons tested on the pis @nodejs/build

rvagg replied:
node-test-binary-arm:

```
case $RUN_SUBSET in
  addons)
    PYTHON=python FLAKY_TESTS=$FLAKY_TESTS_MODE make test-ci-native
    ;;
  *)
    export CC=should-not-compile CXX=should-not-compile # Prevent building
    PYTHON=python FLAKY_TESTS=$FLAKY_TESTS_MODE TEST_CI_ARGS="--run=${RUN_SUBSET},7" make test-ci-js
    ;;
esac
```

i.e. it's used for the addons tests on arm-fanned.

TimothyGu commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+### Options
+
+- `-j <n>`: number of threads used to build the binary or run the parallel 
+  tests
```
Parallel tests are run in parallel by default (by virtue of `-J` specified to the test runner in Makefile), even if the `make` command is `-j1`.

rvagg replied:
There's an awkward catch to this that's not documented anywhere that probably should be since I'm finding Jenkins jobs that don't do this properly: a lot of the -ci targets use this `PARALLEL_ARGS` variable to insert `-j X` for `tools/test.py` (`-J` isn't used in any of the -ci targets). Unfortunately this variable is only set if `$JOBS` exists, and this isn't automatically set just because you supply `-j X` to `make`. In fact, it's kind of difficult to determine the `-j` argument inside a Makefile in a cross-platform way (but we should probably try, there's [an interesting stackoverflow suggestion here](https://stackoverflow.com/questions/5303553/gnu-make-extracting-argument-to-j-within-makefile)). So you have to give `JOBS=X` when running the -ci targets.

`-J` is done automatically for tests on non-ci targets so it takes up "as many cores as you have" for tests as @TimothyGu said.

joyeecheung replied:
@TimothyGu @rvagg Thanks for the correction, I'll add a note in the Makefile about this, and change this part to just "number of threads used to build the binary"

richardlau commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+
+## Makefile
+
+This file is not generated and mantained by humans. Note that this is
```
Typo `maintained`.

Maybe reword this, e.g. `and is maintained` to avoid being misread as `is not (generated and maintained)`.

gibfahn replied:
+1, I totally misread this 😁

richardlau commented on Makefile:
```diff
@@ -811,6 +849,7 @@ $(PKG): release-only
 		--package-path $(MACOSOUTDIR)/pkgs ./$(PKG)
 	SIGN="$(PRODUCTSIGN_CERT)" PKG="$(PKG)" bash tools/osx-productsign.sh
 
+# Builds the MAC OS installer for releases
```
`macOS`

richardlau commented on Makefile:
```diff
@@ -368,6 +379,7 @@ clear-stalled:
 		echo $${PS_OUT} | xargs kill; \
 	fi
 
+# GC test
```
This comment doesn't really convey any additional information than can be inferred from the target name.

richardlau commented on Makefile:
```diff
@@ -100,6 +101,7 @@ install: all
 uninstall:
 	$(PYTHON) tools/install.py $@ '$(DESTDIR)' '$(PREFIX)'
 
+# Clean the artifacts generated by Makefile in most targets
```
I think `clean` is a standard enough makefile target name that this comment shouldn't be necessary. If it is kept, suggest `Remove the artifacts...` reword.

joyeecheung replied:
This is mostly to distinguish `clean` from `docclean`, `test-addons-clean` and friends.

richardlau commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+
+#### Release Targets
+
+- `binary`: builds binary tarballs
```
Maybe a note that this target will fail unless `NODE_VERSION_IS_RELEASE` is #define'd as `1` in `src/node_version.h`.

joyeecheung replied:
@richardlau good catch, though I think this should be documented in the makefile instead.

gibfahn replied:
Would a maintainer need to know that? Seems more relevant for users.

richardlau commented on Makefile:
```diff
@@ -1085,15 +1131,29 @@ lint-addon-docs: test/addons/.docbuildstamp
 	@echo "Please use lint-cpp instead of cpplint"
 
-ifneq ("","$(wildcard tools/eslint/)")
+# tools/eslint is not avialable in the source tarball
```
`available`

rvagg commented on Makefile:
```diff
@@ -234,6 +244,7 @@ test-only: all
 		$(CI_NATIVE_SUITES) \
 		known_issues
 
+# Used by make coverage-test
```
maybe use backticks for \`make coverage-test\`. I read this wrong on first parse and thought it was just awkward wording

rvagg commented on Makefile:
```diff
@@ -408,6 +424,9 @@ test-ci-js: | clear-stalled
 		echo $${PS_OUT} | xargs kill; exit 1; \
 	fi
 
+# Run by most CI, exceptions:
```
I think rpi is the _only_ exception, I've just been grepping through jenkins config files and it's the only one I can find.

rvagg commented on Makefile:
```diff
@@ -584,10 +615,17 @@ docopen: $(apidocs_html)
 	$(RM) -r out/doc
 
+# Prepare the build for running the tests
+# Note that this is not run by the CI jobs on ARM or Linux with FIPS
```
now that I've seen this and noticed `CONFIG_FLAGS` I've gone and edited node-test-commit-linux-fips to not be an exception here, it now does
```
PYTHON=python FLAKY_TESTS=$FLAKY_TESTS_MODE CONFIG_FLAGS=--openssl-fips=$WORKSPACE/fipscanisterdir $MAKE run-ci -j $JOBS
```

so it's using `run-ci` like the rest. I'll do the same for the new jobs I'm building out in node-test-commit-linux-linked that do a similar `./configure` override thing.

So the only thing I'd suggest about this option is that you note that it's called by `run-ci`, we don't actually call it directly anywhere.

_(FYI I also think that node-test-commit-linux-fips wasn't running tests in parallel due to a mis-application of `JOBS` and `-j` and I think that should be fixed now, that would explain why it's been taking more time than it should in some instances).

rvagg commented on Makefile:
```diff
@@ -584,10 +615,17 @@ docopen: $(apidocs_html)
+# Run by most testing CIs, exceptions are:
+#  - ARM (Raspberry Pis): the binaries are cross-compiled elsewhere
+#    before being used to run tests. See test-ci-native.
+#  - Linux with FIPS: the build is done differently, the tests are still
```
fips isn't an exception here, _although_ it does run `test-ci` directly a second time with FIPS enabled

coverage is another exception that doesn't run this

rvagg commented on Makefile:
```diff
@@ -916,6 +957,7 @@ ifeq ($(XZ), 0)
+# Builds the binary tarball for releases
 binary: $(BINARYTAR)
 
 binary-upload: binary
```
you could note that these -upload tasks are strictly for release builds on release machines only, unless you think that's obvious

rvagg commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+
+## configure
+
+To see the help text, run `./configure --help`(unix-like)
```
space needed before `(unix-like)`

gibfahn replied:
I think you can just say `python configure` for both Unix and Windows.

rvagg commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+## configure
+
+To see the help text, run `./configure --help`(unix-like)
+or `.\configure --help` (Windows). Consider updating this file
```
this should probably be `python configure --help`

rvagg commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+- `bench-all`
+- `bench-ci`
+
+### Environment Variables
```
also worth listing:

BUILDTYPE
JOBS
CONFIG_FLAGS
BUILD_DOWNLOAD_FLAGS
BUILD_INTL_FLAGS

rvagg commented:
are you restricting your comment lines shorter than 80 chars?

thanks for doing this btw! I hope we can keep it all in sync as it evolves

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+                        Intl mode: optional local path to icu/ dir, or
+                        path/URL of icu source archive.
+    --download=DOWNLOAD_LIST
+                        Select which packages may be auto-downloaded. valid
```
valid -> Valid

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+                        Download directory [default: ./deps]
+
+  HTTP2:
+    Flags that allows you to control HTTP2 features in Node.js
```
Flags that allows ->  Flags that allow

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+For `deps/v8`:
+
+- `v8`: rebuild `deps/v8` as a git tree, pull the third-party dependencies
+  and builds V8
```
builds -> build ?

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+
+#### Other Test Targets
+
+The following targets uses `node-gyp` bundled inside the npm in the codebase:
```
uses -> use ?

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+- `test-gc`
+
+There are many `test-*` targets that can be run to test a specific part of
+the codebase, most of them using `tools/test.py`. There are also:
```
There are also -> There is also ? (for now)

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+On Windows:
+
+- vcbuild.bat: A Windows Batch Script that locates build tools, provides
+  a subset of the targets avilable in the [Makefile](#makefile), and a few
```
avilable -> available

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+- Makefile: A makefile that can be run with BSD make or GNU make. It
+  provides a set of targets that build and test the Node.js
+  binary, produce releases and documentation, and interact with the
+  CI to run benchamarks or tests. For a detailed guide on this file,
```
benchamarks -> benchmarks

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+                        alternative lib name to link to [default: uv]
+    --shared-libuv-libpath=SHARED_LIBUV_LIBPATH
+                        a directory to search for the shared libuv DLL
+    --shared-openssl    link to a shared OpenSSl DLL instead of static linking
```
OpenSSl -> OpenSSL

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+### Targets
+
+For the targets listed below that do not have detailed descriptions
+in this docuement, please see their comments in the makefile instead.
```
docuement -> document

vsemozhetbyt commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+For the targets listed below that do not have detailed descriptions
+in this docuement, please see their comments in the makefile instead.
+
+Here is a list of targets that will commonly triggered by its users:
```
triggered -> be triggered

richardlau commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+Options:
+  -h, --help            show this help message and exit
+  --prefix=PREFIX       select the install prefix [default: /usr/local]
+  --coverage            Build node with code coverage enabled
```
I think these come direct from the `configure` help, so any changes should be made there.

joyeecheung replied:
@richardlau  Yes.

@vsemozhetbyt Do you want to open a PR to fix the configure script? Or I can do that as well.

vsemozhetbyt replied:
Though these are rather simple fixes, I would rather abstain as I know neither Python nor the format of this file. Sorry(

gibfahn commented on doc/guides/maintaining-the-build-files.md:
````diff
@@ -0,0 +1,386 @@
+    --debug-nghttp2     build nghttp2 with DEBUGBUILD (default is false)
+```
+
+## Makefile
````
I was compiling Alacritty the other day, and I noticed that their `make` actually defaults to showing help (alias for `make help`) rather than doing anything. It actually looks really nice, and it'd be really useful if we could do something similar.

I think a help option to make would be less likely to bitrot than separate documentation.

https://github.com/jwilm/alacritty/blob/245a80078180acc2a0a1addc569c15b77991c1c3/Makefile#L18-L21

gibfahn commented:
>Personally I prefer to leave most of the actual documentation in the code, whiling leave some pointers and necessary explanations in the guide so it's easier to get started with maintaining them.

💯 on this, nothing worse than slightly out-of-date documentation that you can never quite trust.

richardlau commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+                        default compilation environment
+  --enable-vtune-profiling
+                        Enable profiling support for Intel VTune profiler to
+                        profileJavaScript code executed in nodejs. This
```
https://github.com/nodejs/node/pull/17094

richardlau commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,386 @@
+                        architectures.
+  --link-module=LINKED_MODULE
+                        Path to a JS file to be bundled in the binary as a
+                        builtin.This module will be referenced by path without
```
https://github.com/nodejs/node/pull/17094

joyeecheung commented on Makefile:
```diff
@@ -447,29 +466,34 @@ test-node-inspect: $(NODE_EXE)
 test-known-issues: all
 	$(PYTHON) tools/test.py known_issues
 
+# Used by the node-test-npm CI job
```
Hmm I think it's obvious without saying that all the targets here can be run manually (some needs certain environment variables though)

joyeecheung commented:
I think I've addressed most comments. @rvagg @TimothyGu @richardlau @vsemozhetbyt @gibfahn PTAL

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11511/

joyeecheung commented:
Ping @nodejs/build @rvagg @gibfahn @richardlau

joyeecheung commented:
Rebased. New CI: https://ci.nodejs.org/job/node-test-pull-request/11548/

richardlau commented on Makefile:
```diff
@@ -1101,15 +1142,29 @@ lint-addon-docs: test/addons/.docbuildstamp
+lint-ci: lint
+else
+# Inside the git repository
+# This runs linters ond different type of files. It's supposed to
```
typo: `on`?

Trott commented on Makefile:
```diff
@@ -201,10 +205,15 @@ ifeq (,$(wildcard out/$(BUILDTYPE)/cctest))
 endif
 	@out/$(BUILDTYPE)/cctest --gtest_list_tests
 
+# Rebuilds deps/v8 as a git tree, pulls the third-party dependencies
```
Tiny nit and probably a bit onerous, but can we put the directories in this doc in backticks? So `deps/v8` instead of deps/v8? Similarly elsewhere.

Trott replied:
WHoops! Ignore! I thought I was looking at a markdown file. Sorry.

refack commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,194 @@
+
+## vcbuild.bat
+
+To see the help text, run `.\vcbuild help`. Consider updating this file
```
IMHO it should be `vcbuild help`. The `.\` prefix is not idiomatic on windows. Also I'd rephrase to
```markdown
A Windows batch file that encapsulates the steps needed to build & test node on Windows. For available options run `vcbuild help`. Common use is `vcbuild release x64 test`.
Consider updating this file when you need to update the build and testing process on Windows.
```

richardlau replied:
IIRC we settled on the `.\` prefix as it worked across `cmd.exe` and PowerShell.

refack commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,194 @@
+
+On unix-like systems:
+
+- `configure`: A (mostly) Python 2 script that accepts a set of options,
```
I'd move this to it's own sub-section, since _it is_ run on Windows (it's just mostly indirectly).
```markdown
`configure`: A Python script that detects system capabilities and runs [GYP][]. The scripts generate `config.gypi` which includes parameters used by GYP to create system dependant build files. Output is usually in one this formats: Makefile, MSbuild, ninja, or XCode. For a detailed guide on this script, see [configure](#configure).


[GYP]: https://gyp.gsrc.io/docs/UserDocumentation.md
```

joyeecheung commented:
Updated the description of configure as @refack suggested, I've left out the `.\` prefix for the same reason mentioned by @richardlau in https://github.com/nodejs/node/pull/16975#discussion_r152362198. Also improved the capitalization of `Make`.

@refack PTAL, thanks!

bnoordhuis commented on Makefile:
```diff
@@ -62,6 +62,7 @@ V ?= 1
 
 # BUILDTYPE=Debug builds both release and debug builds. If you want to compile
 # just the debug build, run `make -C out BUILDTYPE=Debug` instead.
+# Note that this is for developers, it is not run on the CI. See run-ci.
```
At least the ubuntu1604_sharedlibs_debug_x64 job seems to use it.

bnoordhuis commented on Makefile:
```diff
@@ -385,13 +395,15 @@ CI_JS_SUITES ?= default
 CI_DOC := doctool
 
 # Build and test addons without building anything else
+# On the CI a subset of ARM (Raspberry Pis) runs this, se run-ci
```
s/se/see/, also on line 406.

bnoordhuis commented on Makefile:
```diff
@@ -547,6 +581,7 @@ doc-only: $(apidoc_dirs) $(apiassets)
 	fi;
 	@$(MAKE) -s $(apidocs_html) $(apidocs_json)
 
+# Builds the documetation with the locally built Node.js binary
```
s/documetation/documentation/

bnoordhuis commented on Makefile:
```diff
@@ -1042,6 +1081,7 @@ lint-js:
 jslint: lint-js
 	@echo "Please use lint-js instead of jslint"
 
+# On the CI the output would be emitted in the tap format
```
'Would be' or 'is'?  Also, s/the tap/TAP/

bnoordhuis commented on Makefile:
```diff
@@ -1101,15 +1142,29 @@ lint-addon-docs: test/addons/.docbuildstamp
+
 CONFLICT_RE=^>>>>>>> [0-9A-Fa-f]+|^<<<<<<< [A-Za-z]+
+
+# This is run by the node-test-linter CI job
```
General observation: the new comments should be properly punctuated (dot at the end.)

bnoordhuis commented on Makefile:
```diff
@@ -1205,5 +1253,4 @@ lint-clean:
   uninstall \
-  v8 \
-  website-upload
+  v8
```
Should this have been a separate commit?  It looks like an unrelated change.

joyeecheung replied:
@bnoordhuis I did this when documenting the `-upload` targets and this one does not seem valid any more. I could separate this one out in another commit (there are actually quite a few invalid phony targets)

bnoordhuis commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,199 @@
+On Windows:
+
+- `vcbuild.bat`: A Windows Batch Script that locates build tools, provides
+  a subset of the targets avilable in the [Makefile](#makefile), and a few
```
s/avilable/available/

bnoordhuis commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,199 @@
+  in one of these formats: Makefile, MSbuild, ninja, or XCode project files.
+  (Note: the main Makefile mentioned below is maintained separately by humans).
+  For a detailed guide on this script, see [configure](#configure).
+- `Makefile`: A Makefile that can be run with BSD Make or GNU Make. It
```
Only GNU Make.  There is a BSDMakefile in the top-level directory but it just invokes `gmake`.

bnoordhuis commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,199 @@
+The normal testing process does not run tests of third-party libraries
+inside `deps`, because most of them can run their tests on their own. However,
+there are a few targets in the Makefile that are used to test V8 and npm,
+because we have our own floating patches (not merged in the upstream) that
```
'the upstream tree' (or repository)?

bnoordhuis commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,199 @@
+
+#### Documentation Targets
+
+The Node.js API documentation is built within this codebase. Consider updating
```
IMO, it's not entirely clear what 'within this codebase' means.  Maybe write this as "... built from the Markdown files in `doc/`."

gibfahn commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,199 @@
+- `LOGLEVEL`
+- `PREFIX`
+- `PYTHON`
+- `TEST_CI_ARGS`
```
I'm not sure how useful this is without information about what the variables actually do, and that info would make more sense (to me) in a comment in the Makefile itself.

gibfahn commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,199 @@
+- `configure`: A Python 2 script that detects system capabilities and runs
+  [GYP][]. It generates `config.gypi` which includes parameters used
+  by GYP to create platform-dependent build files. Its output is usually
+  in one of these formats: Makefile, MSbuild, ninja, or XCode project files.
```
When you say `Its output`, do you mean `configure`'s output or GYP's output? If GYP, maybe move it to its own section, as GYP is used on Windows as well.

gibfahn commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,199 @@
+- `test-v8`: run by the `node-test-commit-v8-linux` CI job
+- `test-v8-benchmarks`
+- `test-v8-intl`
+- `test-v8-all`
```
I'm not sure how useful listing all the targets is, and it seems like something that will get out of date quite quickly, vs a `grep '^[a-zA-Z_-]\+: ' Makefile`.

BridgeAR commented:
Ping @joyeecheung

joyeecheung commented:
@BridgeAR ACK, I'll find some time during the weekend to follow up..

joyeecheung commented:
Rebased and updated:

- Use the `Related CI job: ` format to document related CI jobs in the makefile. Removes the `this is run by developers, not the CI...` type of comments because new CI jobs can be added and invalidate that.
- Use the new `##` syntax to document important targets so they show up in `make help`
- Removes the targets and environment variables section in the guide, those should be better documented in the makefile directly instead.

@bnoordhuis @gibfahn PTAL, thanks!

gibfahn commented:
So this changes the output of `make help` from:

![image](https://user-images.githubusercontent.com/15943089/34523278-b38f13f0-f08e-11e7-837c-25e8a60a42da.png)

to 

![image](https://user-images.githubusercontent.com/15943089/34523290-bce371e4-f08e-11e7-821b-d8beb78817a6.png)

When going through the original list, I tried to keep it to just the things I thought new contributors would need (i.e. people who aren't comfortable reading the Makefile). My worry is that by adding a large number of targets we overwhelm newcomers.

I guess the options are:
  1. It's fine, the sorting of the options makes it clear which are which.
  2. We need a `make help` and a `make help-all` (and use `### ` for things that should only be in `help-all`)
  3. We should just put the basic options in `make help`, if people want more info they can just read the Makefile.

For example given those options, I'd be unclear about whether I should use `make doc` or `make doc-only` to build the docs.

joyeecheung commented:
@gibfahn Yes, I agree on shortening the list to make it more friendly to new contributors. I reviewed the list again and I think only `doc-only` and `test-doc` should be added to the list for people trying to contribute to doc. I'll make the changes accordingly. Ideally we can group those targets and indent different subgroups to make them more readable but that can be done in another PR.

joyeecheung commented:
Also, I think `bench` should be removed from the list, see https://github.com/nodejs/node/issues/17053

gibfahn commented:
>Also, I think bench should be removed from the list, see #17053 

Sounds reasonable to me

gibfahn commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,50 @@
+
+This file is not generated, it is maintained by humans. Note that this is not
+usually run on Windows, where [vcbuild.bat](#vcbuild.bat) is used instead.
+To see the help text, run `make help`
```
Maybe put this as the first line in this section (rather than the last).

joyeecheung commented on doc/guides/maintaining-the-build-files.md:
```diff
@@ -0,0 +1,50 @@
+There are three main build files that may be directly run when building Node.js:
+
+- `configure`: A Python 2 script that detects system capabilities and runs
+  [GYP][]. It generates `config.gypi` which includes parameters used by GYP to
```
Ooops, deleted by accident, thanks for catching that.

joyeecheung commented:
Rebased. Planning on landing this on Monday if there are no more outstanding reviews. cc @nodejs/collaborators 

CI: https://ci.nodejs.org/job/node-test-pull-request/12435/

joyeecheung commented:
Landed in c91bd2f24f..89d865715a, thanks!

MylesBorins commented:
Should this land? It will need a manual backport to 6.x and 8.x

joyeecheung commented:
@MylesBorins it's just some doc changes so I would say nope
